sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/ShardReadersPool.java

```diff
@@ -149,10 +157,20 @@ private void readLoop(ShardRecordsIterator shardRecordsIterator) {
           recordsQueue.put(kinesisRecord);
           numberOfRecordsInAQueueByShard.get(kinesisRecord.getShardId()).incrementAndGet();
         }
+        rateLimiter.onSuccess(kinesisRecords);
+      } catch (KinesisClientThrottledException e) {
+        try {
+          rateLimiter.onThrottle(e);
+        } catch (InterruptedException ex) {
+          LOG.warn("Thread was interrupted, finishing the read loop", ex);
+          Thread.currentThread().interrupt();
```

Comment:
uhmmm, we catch an interrupted exception, and then throw out another one. Any reasons why we want to interrupt this thread?

Reply:
Thread.interrupt() does not interrupt the current thread, it sets the thread's interrupt status. The short answer is that catching InterruptedException clears the interrupt status so in order to handle it properly you should always either rethrow the exception or reset the thread's interrupt status so any code higher up the call stack can still detect the interruption. There is a good article by Brian Goetz that goes into the details: https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html.